library/src/main/java/com/wultra/android/mtokensdk/operation/expiration/OperationExpirationWatcher.kt

```diff
@@ -0,0 +1,224 @@
+                    operationsToWatch.removeAll { current -> operations.any { toRemove -> toRemove.equals(current) } }
+                    Logger.d("OperationExpirationWatcher: Stopped watching ${operations.count()} operations.")
+                } else {
+                    operationsToWatch.clear()
+                    Logger.d("OperationExpirationWatcher: Stopped watching all operations.")
+                }
+                prepareTimer()
+            }
+
+            return@synchronized operationsToWatch
+        }
+    }
+
+    private fun prepareTimer() {
+
+        // stop the previous timer
+        timer?.cancel()
+        timer = null
+
+        if (operationsToWatch.isEmpty()) {
+            Logger.d("OperationExpirationWatcher: No operations to watch.")
+            return
+        }
+
+        val firstOp = operationsToWatch.minBy { it.expires }.let {
+            it ?: return@prepareTimer
+        }
+
+        Handler(Looper.getMainLooper()).post {
+
+            // This is a precaution when you'll receive an expired operation from the backend over and over again
+            // and it would lead to infinite refresh time. This also helps when device and backend time is out of sync heavily.
+            // This leads to a minimal "expire report time" of 5 seconds.
+            val interval = max(5, firstOp.expires.toEpochSecond() - currentDateProvider.getCurrentDate().toEpochSecond())
+
+            Logger.d("OperationExpirationWatcher: Scheduling operation expire check in ${interval.toInt()} seconds.")
+
+            val t = Timer("OperationsExpirationWatcherTimer", false)
+            t.schedule(object : TimerTask() {
+                override fun run() {
+                    synchronized(mutex) {
+                        val currentDate = currentDateProvider.getCurrentDate()
+                        val expiredOps = operationsToWatch.filter { it.isExpired(currentDate) }
+
+                        if (expiredOps.isEmpty()) {
+                            return@synchronized
+                        }
+
+                        operationsToWatch.removeAll { it.isExpired(currentDate) }
+                        prepareTimer()
+                        Handler(Looper.getMainLooper()).post {
+                            Logger.d("OperationExpirationWatcher: Reporting ${expiredOps.count()} expired operations.")
+                            listener?.operationsExpired(expiredOps)
+                        }
+                    }
+                }
+            }, interval * 1000)
+            timer = t
+        }
+    }
+}
+
+/**
+ * Interface for listener which gets called when operation expires
+ */
+interface OperationExpirationWatcherListener {
+    /**
+     * Called when operation(s) expire(s).
+     * The method is called on the main thread by the `OperationExpirationWatcher`.
+     * @param expiredOperations array of operations that expired
+     */
+    fun operationsExpired(expiredOperations: List<ExpirableOperation>)
```

Comment:
mark with @MainThread to tell it's delivered on the main thread.
Threading and callback delivery is important on Android.